File: Queue/Queue.rb
load '../Miscellaneous/Node.rb'

class NodeQueue
    def head
        @head
    end
    
    def tail
        @tail
    end
    
    def empty?
        return @head.nil?
    end
    
    def peek
        return nil if empty?
        return @head.data
    end
    
    def enqueue(data)
        node = Node.new(data)
        @tail.next = node unless @tail.nil?
        @head = @tail = node if @head.nil?
        @tail = @tail.next
    end
    
    def dequeue
        return nil if empty?
        
        data = @head.data
        @head = @head.next
        @tail = nil if @head.nil?
        
        return data
    end
end

File: Queue/Queue.py
import collections

class Queue:
    def __init__(self):
        self.__data = collections.deque()

    def isempty(self):
        return len(self.__data) == 0

    def search(self, key):
        return self.__data.index(key) if key in self.__data else None

    def peek(self):
        return self.__data[0] if not self.isempty() else None

    def enqueue(self, data):
        self.__data.append(data)

    def dequeue(self):
        return self.__data.popleft() if not self.isempty() else None
